### DISCUSSION TITLE
Best way to include dynamic data at the sidebar

### DISCUSSION THREAD

**Discussion by luctus** [bridgetownrb/bridgetown#463] [Q&A] [answered]
Hello there,

It's me again, looking for more guidance. This time, it's related to build performance/cache. 

I added a sidebar to every single page that lists ten random featured articles, all pages should list the same 10 random articles, but update them everytime that I re-build the website.

These are the related (simplified) files:

```erb
<!-- _layouts/default.erb -->
<body>
  <main>
    <%= yield %>
  </main>
  <%= render 'sidebar_random_articles' %>
</body>
```

and that partial looks like this:
```erb
<!-- _partials/sidebar_random_articles.erb -->
<aside>
  <% site.collections.articles.resources.select{|s| s.data[:trending] == true}.shuffle(random: Random.new(Date.today.day)).first(10).each do |article| %>
    <div><%= article.data.title %></div>
  <% end %>
</aside>
```

It works totally fine, but it now takes like 50 seconds to build the website (everytime I change something in the code). If I remove that partial, the build takes like 5 seconds.

Do you think is there a way to use the Cache API here in the templates?

Thanks in advance!

**Comment by ayushn21** [comment, marked as answer]
Hey @luctus, I'd recommend extracting the logic to generate the random articles into a `Builder` class and then use that to cache it or maybe even just hold it in memory as it only needs to last for the duration of the build.

Perhaps the builder could add a list of articles to `site.data` and in your template you could do something like:

```
site.data.featured_articles.each { ... }
```

Does that make sense?

**Reply by luctus** [reply]
Lovely. Thank you for the super ultra fast answer! 

I already have a builder, where I load all the resources from an external API. So, you say that I could add something like this?

```rb
class Builders::LoadDataFromApi < SiteBuilder
  def build

    get "https://myapi.com" do |data|
      data[:articles].each do |article|
        add_resource :articles, "#{article[:slug]}.erb" do
          ___ article
        end
      end
    end
    
    site.data.featured_articles = site.collections.articles.resources.select{|s| s.data[:trending] == true}.shuffle.first(10)
  end
end

**Reply by ayushn21** [reply]
Yes, I think in theory that "should" work in theory but I'm not sure on the syntax. Give it a go and see what happens. If you get stuck, are you able to share your repo? I've not done this exact this before and the API's changed a bit recently so will need to poke and prod it a bit to see what works!

**Reply by luctus** [reply]
It worked! I had to use the generator block, like this:
```rb
class Builders::LoadDataFromApi < SiteBuilder
  def build
    get "https://myapi.com" do |data|
      data[:articles].each do |article|
        add_resource :articles, "#{article[:slug]}.erb" do
          ___ article
        end
      end
    end
    generator do
      site.data[:featured_articles] = site.collections.articles.resources.select{|s| s.data[:trending] == true}.shuffle.first(10)
    end
  end
end
```
And then, in the template, I did this:
```erb
<!-- _partials/sidebar_random_articles.erb -->
<aside>
  <% site.data.trending_articles.each do |article| %>
    <div><%= article.data.title %></div>
  <% end %>
</aside>
```
The build time is now 27s. It's definitely better than the previous 50s but still a lot compared to the previous 5s... Is there any other recommendation out there?

Thanks a lot.

**Reply by jaredcwhite** [reply]
How many trending articles are being shuffled? Hard to imagine jumping from 5s to 27s unless we're talking thousands and thousands of articles…

**Reply by luctus** [reply]
Hey @jaredcwhite! I have almost 1000 articles and 30 of them are marked as trending.
I reckon the cost here is related to render that partial 1000 times (it’s present in every article’s page)

**Reply by jaredcwhite** [reply]
So removing or adding that partial causes a big jump in build time?

**Reply by luctus** [reply]
Yep! That’s correct!

**Reply by jaredcwhite** [reply]
OK, good to know…something we can look into optimizing.

In the meantime, you could try memoizing the partial output, effectively cutting it down to 1 iteration:

```erb
<%= site.data.cached_my_partial ||= render("my_partial") %>
```

**Reply by luctus** [reply]
Amazing! it's 8.5 seconds now! thanks a lot!

**Reply by jaredcwhite** [reply]
Cool! Like any caching strategy, it becomes challenging if there's some part of the markup that is article-dependent, but even there various workaround can emerge…

**Reply by luctus** [reply]
Totally agree. At least in my case, having this example of caching a partial is more than enough to start using it. Thanks a lot!
